### PR TITLE
LibJS: Add some AsyncGenerator infrastructure

### PR DESCRIPTION
**Pull request by linusg**

### CODE REVIEW

**Review by davidot — COMMENTED** [review]
LGTM just one small comment, and one thing which doesn't matter (yet or ever)

File: Userland/Libraries/LibJS/Runtime/AsyncGenerator.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    Optional<State> const& async_generator_state() const { return m_async_generator_state; }
+    ExecutionContext const* async_generator_context() const { return m_async_generator_context; }
+    Vector<AsyncGeneratorRequest> const& async_generator_queue() const { return m_async_generator_queue; }
+    Optional<String> const& generator_brand() const { return m_generator_brand; }
+
+private:
+    Optional<State> m_async_generator_state;                 // [[AsyncGeneratorState]]
+    ExecutionContext* m_async_generator_context { nullptr }; // [[AsyncGeneratorContext]]
+    Vector<AsyncGeneratorRequest> m_async_generator_queue;   // [[AsyncGeneratorQueue]]
```
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:41]: Good point, fixed!

File: Userland/Libraries/LibJS/Runtime/AsyncGenerator.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    virtual ~AsyncGenerator() override = default;
+
+    Optional<State> const& async_generator_state() const { return m_async_generator_state; }
+    ExecutionContext const* async_generator_context() const { return m_async_generator_context; }
+    Vector<AsyncGeneratorRequest> const& async_generator_queue() const { return m_async_generator_queue; }
+    Optional<String> const& generator_brand() const { return m_generator_brand; }
+
+private:
+    Optional<State> m_async_generator_state;                 // [[AsyncGeneratorState]]
+    ExecutionContext* m_async_generator_context { nullptr }; // [[AsyncGeneratorContext]]
```
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: Yeah, this is tricky. Here's what the spec wants to do (https://tc39.es/ecma262/#sec-asyncgeneratorstart):

> 2. Let genContext be the [running execution context](https://tc39.es/ecma262/#running-execution-context).
> 3. Set the Generator component of genContext to generator.
> 4. Set the code evaluation state of genContext such that when evaluation is resumed for that [execution context](https://tc39.es/ecma262/#sec-execution-contexts) the following steps will be performed:
>    a-c. [...]
>    d. Remove genContext from the [execution context stack](https://tc39.es/ecma262/#execution-context-stack) and restore the [execution context](https://tc39.es/ecma262/#sec-execution-contexts) that is at the top of the [execution context stack](https://tc39.es/ecma262/#execution-context-stack) as the [running execution context](https://tc39.es/ecma262/#running-execution-context).
> 5. Set generator.[[AsyncGeneratorContext]] to genContext.

So it's sort of both a pointer/reference and a copied value - perhaps we could use a variant? Or copy it and hope that no one mutates any relevant state of the active one before we take it off the stack?

Curious what others think.

[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: Hmm, none of those steps seem like they take genContext by value to me, but I might be misunderstanding something

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: 4d is the important one. We're supposed to somehow keep an execution context that is currently on the C++ stack alive to later restore it.

[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: Right, that's essentially a `vm.pop_execution_context()`

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: hmm, how does that help? `vm.pop_execution_context()` simply removes the pointer from the vector, and once we exit the function where the execution context was created on the stack, it's lost forever -> therefore, keep a pointer, and at some point make a copy that we can push at a later point.

(It's not like I have this all planned in my head, but currently to me that seems like the right way to go)

[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h:40]: Ah I see, I thought we could simply pop it and re-push the existing pointer to the stack, but I guess the lifetime semantics of the pushed execution contexts are not known?

[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/AsyncGenerator.h]: Yep - I'll just go with this for now and we can figure out the details later :)

**Pull request merged**